### DISCUSSION TITLE
Volume calculation between models with different topography

### DISCUSSION THREAD

**Discussion by dmartinezdi** [gempy-project/gempy#896] [General]
**What is your question?**
Hi, I am trying to compare two different models that have the same data but different topography and I am trying to calculate the volume difference between them, but the values that I got from 'raw_arrays' are the same I think because it took the same data to build the model, is there anyway to get the values with the topography cut to be able to measure the volume?

**Comment by javoha** [comment]
Hi,
this seems to be a little hidden in gempy and as far as I know this is not an implemented function. But you can access the vertices of the topogrpahy grid using `geo_model.grid.topography.values`. This might allow you to mask the solution from raw_arrays to get the clipped result. 

Probably more straightforward: `geo_model.grid.topography.topography_mask` returns the mask that can directly be applied to the block solution (e.g. the solution in shape of your resolution).
Let me know if this is what you were looking for or if I got something wrong.
Cheers
Jan

Small edit: I will convert this issue to a discussion for now as it does not strike me as an issue or bug.

**Reply by dmartinezdi** [reply]
Hi Thanks for your answer, I have been trying to apply the mask to the lith block but I have not been able to do it successfully, do you know which would it be the best way to do it?

**Reply by javoha** [reply]
Ok so something along these lines does the trick for me. Sorry for the super dirty code:

```
import numpy.ma as ma
import matplotlib.pyplot as plt

mx = ma.masked_array(geo_model.grid.regular_grid.values[:, 0], mask=geo_model.grid.topography.topography_mask.flatten())
my = ma.masked_array(geo_model.grid.regular_grid.values[:, 1], mask=geo_model.grid.topography.topography_mask.flatten())
mz = ma.masked_array(geo_model.grid.regular_grid.values[:, 2], mask=geo_model.grid.topography.topography_mask.flatten())

mc = ma.masked_array(geo_model.solutions.raw_arrays.lith_block, geo_model.grid.topography.topography_mask.flatten())

m_xyz = np.swapaxes(np.array([mx, my, mz]), 0, 1)

fig = plt.figure()
ax = fig.add_subplot(projection='3d')

ax.scatter(m_xyz[:,0], m_xyz[:,1], m_xyz[:,2], c=mc, cmap="viridis")
```
Is that what you are looking for?

**Reply by dmartinezdi** [reply]
Sort of, I am trying to measure the volume difference between two models, to make the comparison between models and measure how much of cut and values of lithology could be able to extract. I tried to convert the model as blocks to make it in the regular way but I am not being able to do it.

**Reply by javoha** [reply]
I see - guetting the amount of voxels per structural element for each model and estimating the volume by calculating block volumes based on grid resolution is not good enough?

**Reply by dmartinezdi** [reply]
That sounds exactly what I am looking for, but I haven't been able to do it. Can you give me a guide to do it? Also I have the question when I take the voxels out in that way are they cutted by the topography grid?

**Reply by javoha** [reply]
Here is what I came up with. Again, sorry for the rather quick coding job but I hope it does what you need. Basically we create two models with random topographies, calculate the size of a volume block based on extent and resolution of the models, then we mask the resulting solutions by topography and the desired structural element and finally calculate the total volumes to compare. 
Surely not perfect, but I think it should work pretty well. Increasing the resolution will get better results but also increase runtime of course.

[Volume_issue.zip](https://github.com/gempy-project/gempy/files/15429952/Volume_issue.zip)

Let me know if that works for you!

**Reply by dmartinezdi** [reply]
@javoha Thank you so much, that was exactly what I was looking for, I just have a little question, which is the order when I change the lithology, It would be like older as 1 and younger as 3 or on the other way? And 1 would be like the basement right?

**Reply by javoha** [reply]
Hi, glad that it worked. In general it should be young to old, so a value of lith_block = 1 is the top element (youngest), lith_block = 3 in the given example would be the basement. I am not entirely sure if faults would change this but I think they should not. Here also another quick and dirty way to check:

```
fig = plt.figure()
ax = fig.add_subplot(projection='3d')

p= ax.scatter(model_1.grid.regular_grid.values[:,0], 
           model_1.grid.regular_grid.values[:,1], 
           model_1.grid.regular_grid.values[:,2], 
           c=model_1.solutions.raw_arrays.lith_block,
      )

fig.colorbar(p)
```
Cheers,
Jan

**Reply by dmartinezdi** [reply]
Thanks Jan for all your help with this! 

I have one more question but it is not related to this, I am doing a model where I have some Quaternary deposits, thus they are not regular through all the area, do you know any way to model this easily? Right now I am trying to add some points that are above the topography to correct the surfaces but I am not getting everything right.

Cheers, 
David

**Reply by javoha** [reply]
Hi
sorry I somehow missed your new question: Yes, this might be tricky to be honest. If you are able (allowed) to, feel free to send me your model/code and I could take a look. Adding points manually above topography is the best gempy option I can think of. Alternatively you could try adding these structures manually after computing the gempy model. But it depends very much on the exact shapes we are looking at, so I would need to take a look. 
Maybe also making it an onlap relationship might help, but again - hard to say without a more detailed look.

**Reply by dmartinezdi** [reply]
Hi @javoha no worries, thanks for your reply. 
Yeah sure I can send it to you it but I would need an email address to be able to send it, I am going to try the onlap relationship but It will be amazing if you could take a look on it. 
Thanks

**Reply by dmartinezdi** [reply]
Hi @javoha Jan I have sent you the email I do not know if you have seen it and had the time to check it, I am just wondering if it get through or if it might have gone to the spam or something.
Once again thank you,
David

**Comment by javoha** [comment]
Hi, sorry for the silence - I got your model and took a quick look but I didn't have the time to play around with some options. I will let you now as soon as I have tried things.

**Reply by dmartinezdi** [reply]
Hi Jan, no worries, thank you for taking time to look at it, I was just checking if the mail have gone through.

**Reply by javoha** [reply]
Hi,
so I was able to take a look: First, regarding the holes that sometimes appear in the meshes - These are artefacts of the dual contouring. We are aware of them and are working on a solution. It does not change your model though, as the underlying scalar field is the same, it is a bug purely in the mesh extraction step for visualization.
Regarding the bigger picture : I actually cannot run the model with your huge topography tif. What machine are you working on and how long didi it run? From the precomputed solution you send me I can see the issue. Have you tried to calculate the model without topography in higher resolution to identify areas where it does not fit your expectation? 
Cheers
Jan

**Reply by dmartinezdi** [reply]
Hi Jan, 

Oh okay okay I got it, I was trying to extract the mesh surface to make some calculations individually but I have not been able to do it successfully as I was trying to extract them with gemgis but I got something without the correct xyz coordinates, do you happen to have any idea how I can extract them with them? 

Oh god, I can send you a less detailed topography if you like, I am running the model in a Thinkpad e14 with a i7 1165g7 16 gb ram and it takes like 3 min and a half to run with the topography. well the thing to calculate it without the topograpy is that I am not being able to check how the Quaternary behaves as it is not cut by the topography itself. Regarding the volume thing that you helped me through I was wondering if you might know a way to calculate it only inside a polygon that I have in a shp file?

Cheers
David

**Reply by javoha** [reply]
Ah yes, so extracting the meshes works, but you need to back-transform the coordinates of the vertices to get the correct coordinates. Check #916 for example. That way you are able to get the meshes to use in other calculations or software.
Regarding the volume by shapefile - I guess you would need to create some kind of mask to extract only the values within. I am not sure what the smartest way to do that would be, but it should be rather straightfoward.

**Reply by dmartinezdi** [reply]
Hi Jan 

Thanks for that I have been able to extract those and generate the surface, and finishing the volume calculation.

Also I have sent you two new raster files to see if those could work to see if you can check the model with those, one of them the pixel size is 5 m and the other one is 10 m. I have attached those to your email.

Thanks again

**Comment by dmartinezdi** [comment]
Hi @javoha  Jan, i have been trying to make the onlap surface but I do not understand very well how to do it, and have not been successful in the matter, would you be able to guide me through that process please?

**Reply by javoha** [reply]
Hi,
yeah I will try. Again I am not 100% sure it solves your issues but let's see:

First we need to recall again how gempy interpolates. Based on the strucutral frame, elements (rock units) within a group can never crosscut as they are interpolated with a single scalar field. GEmpy uses a bottom to top approach, so younger units within a group will generally conformibly lie on top of older  groups. When we have multiple structural groups, younger groups will unconformably cut of older groups. This is indicated by the default structural relation ERODE.
When we want to model an onlap erlationship we need a younger group to NOT erode the older groups - this is a behavior that needs a little trick in gempy. 
I do think the example we have is really helpful: [LINK](https://docs.gempy.org/tutorials/ch1_fundamentals/ch1_4_onlap_relations.html#sphx-glr-download-tutorials-ch1-fundamentals-ch1-4-onlap-relations-py)
Looking at the cross section image:
![image](https://github.com/user-attachments/assets/78729a0d-0632-4f10-b0f0-e21b95294b72)
The normal gempy behavior would lead to the structural group _right_series_ containing _rock1_ and _rock2_ cutting of the older group _left_series_ with _rock3_. In order to get _right_series_ to onlap on _left_series_ we need to change two things:
1) We need to set its strucutral relation to ONLAP, forcing it to not cut older groups
2) As gempy normally goes from bottom to top, but top (younger) now is forced to not cut older (bottom), we need to add the upper bound of our _left_series_. This forces us to add a new structural group with an element we called _onlap_surface_ in this example. Note that itself it still has the ERODE stack relation as it limits the left_series. This element is what our younger group reight series will now onlap on.

Does this make sense? I really hope it does :)

Feel free to ask if anything is unclear.

Cheers,
Jan

**Reply by dmartinezdi** [reply]
Hi @javoha, 

Thank you, that sounds good but I do not understand very well how to setup the onlap relation, so I have this structural frame:
![image](https://github.com/user-attachments/assets/bca33234-b53c-4f24-a594-f174a9ab8ad4)
and I was trying to put it between Kd3 and Tba but I do not get pretty well if I have to had the data as the onlap and then data as an erode version of it? like I would put Tba as the onlap and then again as erode or how should I do that?

Also I have another little question if you happen to know, I was trying to get out the model plots in 3d and 2d with a scale bar is that possible? I saw like some kind of model in gpyviewer but I am not certain how to use it. Is it possible? And, is it possible to extract like the 2d plots as other format as a shp or a csv with the data of the cell or something like that?

Thank you 

Cheers, 
David

**Reply by javoha** [reply]
Hi,
so the onlap relation mostly makes sense if you have multiple elements in a structural group that onlap on a surface. And yes - then you need to define it as an additional surface. If the onlapping group is just a single element, using erode as of usual kind of works the same way. Maybe this ugly sketch helps:
![image](https://github.com/user-attachments/assets/7a10aae4-b302-464a-90ef-371831db5a0a)
Case A is when you need an additional onlap surface (red), as multiple structural elements of the younger group (orange, green) onlap on this surface. In Case B it does not matter as the structure is just filled with a single element and you can just put it to erode. 

Regarding your second question: What do you mean by scale bar. It is possible to plot the scalar field instead of the lithologies. And as you can extract the lithology block you basically have the data per cell as well as the surfaces meshes for 3D plotting wherever you want. Is that what you mean?

Cheers,
Jan
